libparsec/crates/client/src/workspace/transactions/fd_read.rs

```diff
@@ -80,7 +81,7 @@ pub async fn fd_read(
         match found {
             Some(chunk_data) => {
                 chunk
-                    .copy_between_start_and_stop(chunk_data, buf)
+                    .copy_between_start_and_stop(chunk_data, offset, buf, &mut buf_size)
```

Comment:
`offset` is passed as parameter in the function, so I would expect it to be only used for the first chunk and not the subsequent ones :/

Reply:
I tried to keep those changes minimal. The reference logic for building data is demonstrated in the test `Storage`:
```rust
    fn read_chunk(&self, chunk: &Chunk) -> &[u8] {
        let data = self.read_chunk_data(chunk.id);
        let start = (chunk.start - chunk.raw_offset) as usize;
        let stop = (chunk.stop.get() - chunk.raw_offset) as usize;
        data.get(start..stop).unwrap()
    }

    fn build_data(&self, chunks: &[Chunk], size: u64, offset: u64) -> Vec<u8> {
        let start = offset as usize;
        let mut result: Vec<u8> = vec![0; size as usize];
        for chunk in chunks {
            let data = self.read_chunk(chunk);
            result[chunk.start as usize - start..chunk.stop.get() as usize - start]
                .copy_from_slice(data);
        }
        result
    }

    // File operations

    pub fn read(&self, manifest: &LocalFileManifest, size: u64, offset: u64) -> Vec<u8> {
        let (written_size, chunks) = prepare_read(manifest, size, offset);
        let data = self.build_data(&chunks.collect::<Vec<_>>(), written_size, offset);
        p_assert_eq!(data.len() as u64, written_size);
        data
```

The TLDR is that the `written_size` returned by `prepare_read` should be used to allocate a `Vec<u8>` filled with zero. Then, each chunk correspond to a slice of data that should be copied into that buffer.

The issue with `fd_read` is that the buffer is a `std::io::Write`, hence the complicated logic.

I'll let you change it as you see fit.